File: projects/sdk/ts/src/framing.ts
/**
 * Length-prefixed framing for JSON-RPC messages over a stream transport.
 *
 * Wire format (matches projects/sdk/src/framing.rs):
 *   [ 4-byte big-endian uint32 length ][ length bytes of body ]
 */
import type { Readable, Writable } from 'node:stream';

/**
 * Largest frame body either side will read or write. Mirrors the 16 MiB cap
 * in the Rust SDK; both sides MUST agree to prevent malicious or buggy peers
 * from forcing unbounded allocation.
 */
export const MAX_FRAME = 16 * 1024 * 1024;

export class FrameTooLargeError extends Error {
  constructor(size: number) {
    super(`frame too large: ${size} bytes (max ${MAX_FRAME})`);
    this.name = 'FrameTooLargeError';
  }
}

/** Write one framed message: 4-byte length header followed by body. */
export function writeFrame(stream: Writable, body: Uint8Array): Promise<void> {
  if (body.length > MAX_FRAME) {
    return Promise.reject(new FrameTooLargeError(body.length));
  }
  const header = Buffer.alloc(4);
  header.writeUInt32BE(body.length, 0);
  return new Promise((resolve, reject) => {
    stream.write(header, err => {
      if (err) return reject(err);
      if (body.length === 0) return resolve();
      stream.write(body, err2 => (err2 ? reject(err2) : resolve()));
    });
  });
}

/**
 * Reader that pulls one frame at a time from a Node Readable. Internally
 * accumulates bytes and yields complete frames as Buffers.
 */
export class FrameReader {
  private buf: Buffer = Buffer.alloc(0);
  private waiters: Array<(value: Buffer | null) => void> = [];
  private ended = false;
  private error: Error | null = null;

  constructor(stream: Readable) {
    stream.on('data', (chunk: Buffer) => {
      this.buf = (this.buf.length === 0 ? chunk : Buffer.concat([this.buf, chunk])) as Buffer;
      this.drain();
    });
    stream.on('end', () => {
      this.ended = true;
      this.drain();
    });
    stream.on('error', err => {
      this.error = err;
      this.drain();
    });
  }

  /** Resolve to the next complete frame, or null on stream end. Rejects on error. */
  read(): Promise<Buffer | null> {
    return new Promise((resolve, reject) => {
      const tryServe = () => {
        if (this.error) {
          reject(this.error);
          return true;
        }
        if (this.buf.length >= 4) {
          const n = this.buf.readUInt32BE(0);
          if (n > MAX_FRAME) {
            reject(new FrameTooLargeError(n));
            return true;
          }
          if (this.buf.length >= 4 + n) {
            const body = this.buf.subarray(4, 4 + n);
            // Copy out so callers can keep references after we slide forward.
            const out = Buffer.from(body);
            this.buf = this.buf.subarray(4 + n);
            resolve(out);
            return true;
          }
        }
        if (this.ended) {
          resolve(null);
          return true;
        }
        return false;
      };
      if (tryServe()) return;
      this.waiters.push(value => {
        if (value === null) resolve(null);
        else resolve(value);
      });
    });
  }

  private drain(): void {
    while (this.waiters.length > 0) {
      const w = this.waiters[0]!;
      if (this.error) {
        this.waiters.shift();
        // Hand the error back via a rejected wait. We can't reject through
        // the resolver, so re-create the promise pattern: best-effort drop.
        w(null);
        continue;
      }
      if (this.buf.length < 4) {
        if (this.ended) {
          this.waiters.shift();
          w(null);
          continue;
        }
        return;
      }
      const n = this.buf.readUInt32BE(0);
      if (n > MAX_FRAME) {
        // Surface via next read() call; here we just drop the waiter and
        // let read() observe the error on its next invocation.
        return;
      }
      if (this.buf.length < 4 + n) {
        if (this.ended) {
          this.waiters.shift();
          w(null);
          continue;
        }
        return;
      }
      const body = Buffer.from(this.buf.subarray(4, 4 + n));
      this.buf = this.buf.subarray(4 + n);
      this.waiters.shift();
      w(body);
    }
  }
}

File: projects/sdk/rust/src/framing.rs
//! Length-prefixed framing for JSON-RPC messages over a stream transport.
//!
//! Wire format: 4-byte big-endian u32 length followed by exactly that many UTF-8 bytes.

use anyhow::{Result, ensure};
use tokio::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};

/// Maximum frame body size (16 MiB). Guards against malformed/malicious peers.
const MAX_FRAME: u32 = 16 * 1024 * 1024;

/// Write one framed message: length header + body.
pub async fn write_frame<W: AsyncWrite + Unpin>(w: &mut W, body: &[u8]) -> Result<()> {
    let len = u32::try_from(body.len())
        .map_err(|_| anyhow::anyhow!("message too large to frame: {} bytes", body.len()))?;
    w.write_all(&len.to_be_bytes()).await?;
    w.write_all(body).await?;
    Ok(())
}

/// Read one framed message: consume 4-byte length header, then that many bytes.
pub async fn read_frame<R: AsyncRead + Unpin>(r: &mut R) -> Result<Vec<u8>> {
    let mut hdr = [0u8; 4];
    r.read_exact(&mut hdr).await?;
    let len = u32::from_be_bytes(hdr);
    ensure!(len <= MAX_FRAME, "frame too large: {len} bytes");
    let mut buf = vec![0u8; len as usize];
    r.read_exact(&mut buf).await?;
    Ok(buf)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn roundtrip_small_message() {
        let msg = b"hello world";
        let mut buf: Vec<u8> = Vec::new();
        write_frame(&mut buf, msg).await.unwrap();
        // buf = [0,0,0,11] ++ "hello world"
        assert_eq!(buf.len(), 4 + msg.len());
        let got = read_frame(&mut buf.as_slice()).await.unwrap();
        assert_eq!(got, msg);
    }

    #[tokio::test]
    async fn roundtrip_empty_message() {
        let mut buf: Vec<u8> = Vec::new();
        write_frame(&mut buf, b"").await.unwrap();
        let got = read_frame(&mut buf.as_slice()).await.unwrap();
        assert!(got.is_empty());
    }

    #[tokio::test]
    async fn rejects_oversized_frame() {
        // Craft a header claiming MAX_FRAME + 1 bytes
        let big = (MAX_FRAME + 1).to_be_bytes();
        let err = read_frame(&mut big.as_slice()).await;
        assert!(err.is_err());
        assert!(err.unwrap_err().to_string().contains("frame too large"));
    }
}
